### PR TITLE
Delta vault roll to next option

### PR DESCRIPTION
**Pull request by chuddster**

### CODE REVIEW

File: contracts/libraries/GnosisAuction.sol [hunk 1]
```diff
@@ -103,6 +118,58 @@ library GnosisAuction {
...
+    function placeBid(BidDetails memory bidDetails)
+        internal
+        returns (uint256 oTokensToBuy, uint256 bidAmount)
+    {
+        // calculate how much to allocate
+        bidAmount = bidDetails
+            .lockedBalance
+            .mul(bidDetails.optionAllocationPct)
+            .div(100 * 10**2);
+        oTokensToBuy = 0; //bidAmount.div(PREMIUM)
```
Comment (chuddster) [inline contracts/libraries/GnosisAuction.sol:130]: still TBD what premium we are willing to give per oToken. 

should we offer the floor premium of the auction?

File: contracts/libraries/GnosisAuction.sol [hunk 2]
```diff
@@ -103,6 +120,63 @@ library GnosisAuction {
...
 
+    function placeBid(BidDetails memory bidDetails)
+        internal
+        returns (uint256 oTokensToBuy, uint256 bidAmount)
+    {
+        // calculate how much to allocate
+        bidAmount = bidDetails
+            .lockedBalance
+            .mul(bidDetails.optionAllocationPct)
+            .div(100 * 10**2);
```
Comment (kenchangh) [inline contracts/libraries/GnosisAuction.sol:131]: Why do we have to do `100 * 10**2`?

File: contracts/libraries/GnosisAuction.sol [hunk 3]
```diff
@@ -20,6 +20,12 @@ library GnosisAuction {
         address manager
     );
 
+    event PlaceAuctionBid(
+        uint256 auctionId,
+        address auctioningToken,
+        uint256 bidAmount
```
Comment (kenchangh) [inline contracts/libraries/GnosisAuction.sol:26]: Let's add `address bidder`

File: contracts/interfaces/IGnosisAuction.sol [hunk 4]
```diff
@@ -68,4 +68,9 @@ interface IGnosisAuction {
         bytes32[] memory _prevSellOrders,
         bytes calldata allowListCallData
     ) external returns (uint64);
+
+    function claimFromParticipantOrder(
```
Comment (chuddster) [inline contracts/interfaces/IGnosisAuction.sol:72]: bidder actually needs to call [claimFromParticipantOrder](https://github.com/gnosis/ido-contracts/blob/a766a8b30de14781961158b772393cc49b900714/contracts/EasyAuction.sol#L568) as well. so will need to call this on behalf of delta vault in cron job in separate pr

File: contracts/libraries/GnosisAuction.sol [hunk 5]
```diff
@@ -103,6 +123,73 @@ library GnosisAuction {
...
+        );
+
+        // divide the `asset` sellAmount by the target premium per oToken to
+        // get the number of oTokens to buy (8 decimals)
+        buyAmount = uint96(
+            uint256(sellAmount)
+                .mul(10**bidDetails.assetDecimals)
+                .div(bidDetails.optionPremium)
+                .mul(10**8)
+                .div(10**bidDetails.assetDecimals)
```
Comment (chuddster) [inline contracts/libraries/GnosisAuction.sol:148]: multiplying by asset decimals twice - once for bidDetails.optionPremium which has `asset` decimals and once for shifting it to 8 decimals

File: contracts/vaults/RibbonDeltaVault.sol [hunk 6]
```diff
@@ -199,6 +200,7 @@ contract RibbonDeltaVault is RibbonVault, OptionsDeltaVaultStorage {
         bidDetails.lockedBalance = lockedBalance;
         bidDetails.optionAllocationPct = optionAllocationPct;
         bidDetails.optionPremium = optionPremium;
+        bidDetails.bidder = owner();
```
Comment (kenchangh) [inline contracts/vaults/RibbonDeltaVault.sol:206]: This should be `bidDetails.bidder = msg.sender` to save gas

File: contracts/interfaces/IGnosisAuction.sol (hunk 4 above)
Comment (kenchangh) [reply contracts/interfaces/IGnosisAuction.sol:72]: Do we want to make a pass-through function on the Delta Vault that calls `claimFromParticipantOrder`? Feels better to have everything on 1 contract so people can poke it

File: contracts/libraries/GnosisAuction.sol [hunk 7]
```diff
@@ -124,47 +125,55 @@ library GnosisAuction {
...
     {
         // calculate how much to allocate
-        bidAmount = bidDetails
-            .lockedBalance
-            .mul(bidDetails.optionAllocationPct)
-            .div(10000);
+        sellAmount = uint96(
+            bidDetails.lockedBalance.mul(bidDetails.optionAllocationPct).div(
+                10000
+            )
```
Comment (kenchangh) [inline contracts/libraries/GnosisAuction.sol:138]: Check for uint96 overflow

File: contracts/libraries/GnosisAuction.sol [hunk 8]
```diff
@@ -124,47 +125,55 @@ library GnosisAuction {
...
-            bidAmount <= type(uint96).max,
-            "bidAmount > type(uint96) max value!"
+            sellAmount <= type(uint96).max,
+            "sellAmount > type(uint96) max value!"
         );
         require(
-            oTokensToBuy <= type(uint96).max,
-            "oTokensToBuy > type(uint96) max value!"
+            buyAmount <= type(uint96).max,
+            "buyAmount > type(uint96) max value!"
```
Comment (kenchangh) [inline contracts/libraries/GnosisAuction.sol]: Checking for uint96 overflow here doesn't actually do anything. We need to do the checks before casting it to uint96.

This works:

```solidity
uint256 number = someNumber;
require(number <= type(uint96.max));
uint96 smallNumber = uint96(number);
```

This does not work:

```solidity
uint256 number = someNumber;
// casting already truncates the integer to be the size uint96.
uint96 smallNumber = uint96(number); 
require(smallNumber <= type(uint96.max));
```

Comment (chuddster) [reply contracts/libraries/GnosisAuction.sol]: yes good point

File: contracts/vaults/RibbonDeltaVault.sol [hunk 9]
```diff
@@ -152,13 +161,12 @@ contract RibbonDeltaVault is RibbonVault, OptionsDeltaVaultStorage {
 
         address counterpartyNextOption =
             counterpartyThetaVault.optionState().nextOption;
-        require(counterpartyNextOption != address(0));
+        require(counterpartyNextOption != address(0), "!thetavaultclosed");
```
Comment (kenchangh) [inline contracts/vaults/RibbonDeltaVault.sol:164]: nit deltavaultclosed

**Review by kenchangh — APPROVED** [review]

**Pull request merged**